### PR TITLE
Share/Open with newpipe" add enqueue choice for players

### LINKED ISSUE #1487: Queuing background videos outside of app[Bug report?]
Hey NewPipe team,
I have a suggestion/bug report for the background player.When i'm inside the newpipe app and i queue a background video the current video playing in the background continues to play then once finished goes to the next queued item and the newly queued background video just gets added to the end of the queue which is great and whats expected :)

however if I share a video to be played in background by newpipe from a separate app ,the newly queued video interrupts/stops any currently playing background video and plays itself instead of the video originally playing as opposed to just queuing to play after the current video which you would see if en-queued inside the app. Fixing this minor issue would be great :)

Once again thanks NewPipe team!!!

### LINKED ISSUE #4779: Add Enqueue button in "Always Ask" Share Menu
### Checklist

- [x] I checked, but didn't find any duplicates (open OR closed) of this issue in the repo. 
- [x] I have read the contribution guidelines given at https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md.
- [x] This issue contains only one feature request. I will open one issue for every feature I want to request.

#### Describe the feature you want

Until newPipe 0.20.2, while playing a video in background or pop-up mode, sharing another video from others apps to newPipe would enqueue them directly. However this is changing in the upcoming release because it was determined to be bad UX #4562. therefor I am raising this issue to discuss an  alternative for it.

A direct quote from developer:
>When sharing a video to NewPipe and a play button is pressed, replace the current queue instead of enqueueing. Enqueueing can still be achieved opening info and then enqueueing, but enqueueing directly was bad UX since the user would not see what happened.

#### Is your feature request related to a problem? Please describe it

These are my related conversations with the developer about this issue: https://github.com/TeamNewPipe/NewPipe/pull/4562#issuecomment-712496743, https://github.com/TeamNewPipe/NewPipe/pull/4562#issuecomment-719837705

To summarize: when a video is shared, enqueueing to pop-up or background mode was done directly, however Stypox suggested an alternative to direct enqueueing by shareing video to newPipe, then opening its details from menu, then long pressing enqueue button... however this was a tedious alternative as it would take too much time and effort to do so compared to old method, especially if multiple videos are to be enqueued.. (you can try enqueueing 5 video yourself to see the vast difference).

also when sharing to pop-up or background mode, users actually do not really want video details... they actually just wants the video to be enqueued... if the users indeed wanted video details they would have choosen the first option (video details). ..more so, it results in wasting data for loading the video details page for each video. 

#### Additional context

The alternative for old enqueueing method should be as easy as the old method (enqueueing directly)..  it sould avoid wasting to much time and data.

@opusforlife2 suggested when the share menu appears, long pressing on pop-up or background option would enqueue the video ... my suggestion isn't far away from his but more intuitive if that's the right word. Instead of long pressing, a ( + )"plus sign" or something similar is to be added next to both modes on the right. 
Screenshot_20201102-210713

The work done in #4425  "Replace specific enqueue options with one" could be used to inspire more alternatives. 
#### How will you/everyone benefit from this feature?

Makes life a little bit easier... but seriously, try enqueueing 5 video yourself in both ways to see the vast difference.

### PR DESCRIPTION
**Pull request by litetex**
#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
Successor to #5850 (please take a look at that PR for the basic description).

This is a new PR because it's not possible for a maintainer to modify the old one and I otherwise had to request so many tiny changes which would not be easily doable.

TL;DR - changes on top:
* Fixed/Reformatted code
* Refactored UI in the background (nearly looks the same)
* Some minor fixes (theme not working etc)

#### Before/After Screenshots/Screen Record
grafik grafik

https://user-images.githubusercontent.com/40789489/137603434-60662612-2262-4048-95b9-b10c2d4abe2b.mp4

#### Fixes the following issue(s)
- Fixes #4779
- Fixes #1487

#### APK testing 
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

File: app/src/main/res/xml/video_audio_settings.xml [hunk 1]
```diff
@@ -119,6 +119,13 @@
             app:singleLineTitle="false"
             app:iconSpaceReserved="false" />
 
+        <SwitchPreferenceCompat
```
Comment (Stypox) [inline app/src/main/res/xml/video_audio_settings.xml:122]: Do we really need to also add the setting here? Isn't the switch in the menu enough?

File: app/src/main/res/layout/dialog_share_context.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+
+    <RadioGroup
+        android:id="@+id/player_group"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content">
+
+    </RadioGroup>
```
Comment (Stypox) [inline app/src/main/res/layout/dialog_share_context.xml:14]:
```suggestion
        android:layout_height="wrap_content" />
```

File: app/src/main/res/layout/dialog_share_context.xml [hunk 3]
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (Stypox) [inline app/src/main/res/layout/dialog_share_context.xml:2]: Would it be possible to use a `ConstraintLayout` to prevent view nesting?

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java [hunk 4]
```diff
@@ -796,12 +864,37 @@ public void handleChoice(final Choice choice) {
...
+                        NavigationHelper.enqueueOnPlayer(
+                                this,
+                                playQueue,
+                                MainPlayer.PlayerType.AUDIO);
+                    } else if (choice.playerChoice.equals(popupPlayerKey)) {
+                        NavigationHelper.enqueueOnPlayer(
+                                this,
+                                playQueue,
+                                MainPlayer.PlayerType.POPUP);
+                    }
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/RouterActivity.java:889]:
```suggestion
                    MainPlayer.PlayerType playerType = MainPlayer.PlayerType.AUDIO;
                    if (choice.playerChoice.equals(videoPlayerKey)) {
                        playerType = MainPlayer.PlayerType.VIDEO;
                    } else if (choice.playerChoice.equals(popupPlayerKey)) {
                        playerType = MainPlayer.PlayerType.POPUP;
                    } // else background, init value of playerType
                    
                    NavigationHelper.enqueueOnPlayer(this, playQueue, playerType);
```

**Review by Stypox — CHANGES_REQUESTED** [review]
Do we really need to hide the enqueue option under an "Advanced options" menu? The space needed for the single switch is the same as that needed for the advanced options menu, and the amount of data on the screen is also the same, so I wouldn't do it like this. I'd just have the switch there and call it a day: people who don't know what it is or who don't use it can ignore it without any problem, but people who want to use it save one click.

File: app/src/main/res/xml/video_audio_settings.xml (hunk 1 above)
Comment (litetex) [reply app/src/main/res/xml/video_audio_settings.xml:122]: Yes because it's not possible to reach that switch in the share dialog if a default value was chosen (e.g. always open with video player) :)

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java [hunk 5]
```diff
@@ -796,12 +864,29 @@ public void handleChoice(final Choice choice) {
...
+                // If prioritize_enqueue is enabled, then enqueue this stream.
+                // Otherwise, open preferred player instead
+                if (prioritizeEnqueue) {
+                    MainPlayer.PlayerType playerType = MainPlayer.PlayerType.AUDIO;
+                    if (choice.playerChoice.equals(videoPlayerKey)) {
+                        playerType = MainPlayer.PlayerType.VIDEO;
+                    } else if (choice.playerChoice.equals(popupPlayerKey)) {
+                        playerType = MainPlayer.PlayerType.POPUP;
+                    } // else background, init value of playerType
+
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/RouterActivity.java:880]:
```suggestion

```

**Comment by litetex** [comment]
https://github.com/TeamNewPipe/NewPipe/pull/7266#pullrequestreview-812280290

This issue has to go back to the drawing board...
Set it's state to draft.

**Comment by SameenAhnaf** [comment]
https://github.com/TeamNewPipe/NewPipe/issues/1487 could be linked as well.

**Comment by offshade** [comment]
Is there a version with this option available? This is great and much needed.

**Comment by opusforlife2** [comment]
> Is there a version with this option / enqueue available? This is great and much needed.

@offshade While logged in, go to the Checks tab of this PR, then click on Artifacts. You can download the zipped APK from it. It will install a debug version of Newpipe which won't interfere with the main version. You can delete it after the main version gets this feature and resume using that.

**Comment by offshade** [comment]
> > Is there a version with this option / enqueue available? This is great and much needed.
> 
> @offshade While logged in, go to the Checks tab of this PR, then click on Artifacts. You can download the zipped APK from it. It will install a debug version of Newpipe which won't interfere with the main version. You can delete it after the main version gets this feature and resume using that.

This is so awesome! Thank you very very much.
Is there  a thread for this version in case it has updates etc?

**Comment by offshade** [comment]
Version doesn't work anymore.

**Comment by opusforlife2** [comment]
@offshade Wait for the PR to be rebased.

**Comment by offshade** [comment]
> @offshade Wait for the PR to be rebased.

I will wait for anything, although being clueless about technical stuff, I have no idea what you just said... 😂

**Comment by rancidfrog** [comment]
Will this be merged?
This issue has been raised for a while now.
Is it so hard to just have an extra enqueue option in dialog?
I believe in previous iteration it used to be available,
Now when always ask option is enabled it does not appear:

**Comment by litetex** [comment]
> Will this be merged?

In the current state not.

>This issue has been raised for a while now.
Is it so hard to just have an extra enqueue option in dialog?

Read the above comments.

> I believe in previous iteration it used to be available,

No.

**Comment by ThisIsPaulDaily** [comment]
The enqueue option could be the non-default dialog option if you are concerned about background versus foreground player antics. This would force the user to make an informed decision to set always open this way. This has been broken since version 18 or something when background player lost the ability to enqueue separately. 

Also, addressing the history comment as a work around, it isn't a work around because I'll often enqueue multiple videos and they won't be marked in my history unless they play.

**Comment by MD77MD** [comment]
@litetex 

could this integrated in the need release 😄

**Comment by pc00per** [comment]
I'm confused with the status of this PR. Isn't this already worked upon?

**Comment by Stypox** [comment]
Yes, but there is currently no agreement on how the UI should look and behave

**Comment by MD77MD** [comment]
any chance for this to be reviewed.?

**Comment by offshade** [comment]
Yes, please reopen!

**Comment by offshade** [comment]
Ehm, no. I wouldn't know what to do, completely clueless.

**Comment by ThisIsPaulDaily** [comment]
I would really appreciate this getting fixed and it broke the way I used NewPipe.
